### ISSUE TITLE
Output vector is -Inf

### ISSUE THREAD

**Issue by 0phoff** [marvis/pytorch-yolo2#39]
I modified the yolo-voc configs to have a person detector (class_num=1, output filters=30,...) and trained it using the darknet implementation from pjreddie. I now want to switch to using pytorch, because it is easier to modify.

So I ran the detector using your implementation in pytorch, but I get no detections on the input images, and upon closer inspection, I noticed the output vector of the network has -Inf for every value.
I thus looked at the intermediate results of the network, and noticed a part of the output vector after the first convolution,batchnorm,leakyrelu contains NaN values.

Now I'm not too sure what I should do to debug this further...
here is a [link](https://gist.github.com/0phoff/187c57c2a7e4b84f6628483e867f9492) to the config file. I can also provide the weight file if needed.

**Comment by 0phoff** [comment]
I followed the [_how_to_debug_](https://github.com/marvis/pytorch-yolo2/blob/master/how_to_debug.md) guide to print the output of the first layer, here are the first 100 outputs of the first layer. I also played around with the code a bit, and the first NaN values come after the first BatchNorm (sub)layer.

<details><summary>Darknet</summary>

```
0: 5.754759
1: -0.236937
2: -0.236937
3: -0.236937
4: -0.236937
5: -0.236937
6: -0.236937
7: -0.236937
8: -0.236937
9: -0.236937
10: -0.236937
11: -0.236937
12: -0.236937
13: -0.236937
14: -0.236937
15: -0.236937
16: -0.236937
17: -0.236937
18: -0.236937
19: -0.236937
20: -0.236937
21: -0.236937
22: -0.236937
23: -0.236937
24: -0.236937
25: -0.236937
26: -0.236937
27: -0.236937
28: -0.236937
29: -0.236937
30: -0.236937
31: -0.236937
32: -0.236937
33: -0.236937
34: -0.236937
35: -0.236937
36: -0.236937
37: -0.236937
38: -0.236937
39: -0.236937
40: -0.236937
41: -0.236937
42: -0.236937
43: -0.236937
44: -0.236937
45: -0.236937
46: -0.236937
47: -0.236937
48: -0.236937
49: -0.236937
50: -0.236937
51: -0.236937
52: -0.236937
53: -0.236937
54: -0.236937
55: -0.236937
56: -0.236937
57: -0.236937
58: -0.236937
59: -0.236937
60: -0.236937
61: -0.236937
62: -0.236937
63: -0.236937
64: -0.236937
65: -0.236937
66: -0.236937
67: -0.236937
68: -0.236937
69: -0.236937
70: -0.236937
71: -0.236937
72: -0.236937
73: -0.236937
74: -0.236937
75: -0.236937
76: -0.236937
77: -0.236937
78: -0.236937
79: -0.236937
80: -0.236937
81: -0.236937
82: -0.236937
83: -0.236937
84: -0.236937
85: -0.236937
86: -0.236937
87: -0.236937
88: -0.236937
89: -0.236937
90: -0.236937
91: -0.236937
92: -0.236937
93: -0.236937
94: -0.236937
95: -0.236937
96: -0.236937
97: -0.236937
98: -0.236937
99: -0.236937
```
</details>

<details><summary>Pytorch</summary>

```
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
 nan
[torch.cuda.FloatStorage of size 100]
```
</details>

**Comment by 0phoff** [comment]
I tried to debug this problem further, and found out the weights from both networks are different!  
Here is the code I used for printing, and the results.

__Darknet__
```C
/* In examples/detector.c:test_detector() */

//image sized = letterbox_image(im, net.w, net.h);
image sized = im;

layer ll = net.layers[0];
float *Y = ll.weights;
printf("----------\n");
for(j = 0; j < 100; j++) printf("%d: %f\n", j, Y[j]);
printf("\n\n");
```
<details><summary>Darknet Output</summary>

```
----------
0: -0.057812
1: 0.008488
2: 0.027734
3: -0.126044
4: 0.019132
5: 0.090675
6: -0.019783
7: 0.034583
8: 0.025900
9: -0.057418
10: 0.027847
11: 0.059884
12: -0.135119
13: 0.018076
14: 0.097718
15: -0.055052
16: 0.013282
17: 0.024293
18: -0.022654
19: 0.004577
20: 0.019717
21: -0.091178
22: -0.005140
23: 0.053772
24: -0.003668
25: -0.000185
26: 0.000560
27: 0.034911
28: 0.055543
29: 0.061609
30: 0.035788
31: -0.048132
32: -0.047910
33: 0.033296
34: -0.088302
35: -0.052671
36: 0.046841
37: 0.055535
38: 0.066024
39: 0.040595
40: -0.069638
41: -0.067265
42: 0.039733
43: -0.093759
44: -0.052298
45: -0.043797
46: -0.045754
47: -0.025026
48: -0.044076
49: -0.076992
50: -0.057401
51: -0.009686
52: -0.057275
53: -0.018817
54: 0.153752
55: -0.138363
56: -0.051715
57: 0.264345
58: -0.073381
59: -0.192942
60: 0.225416
61: 0.039763
62: -0.176477
63: 0.189306
64: -0.183411
65: -0.093585
66: 0.380824
67: -0.093574
68: -0.316861
69: 0.297664
70: 0.060877
71: -0.247239
72: 0.077353
73: -0.076158
74: -0.030189
75: 0.131463
76: -0.018913
77: -0.120964
78: 0.084208
79: 0.047146
80: -0.117478
81: -0.002414
82: -0.001510
83: 0.001867
84: -0.002610
85: -0.000096
86: 0.003090
87: -0.001463
88: 0.001551
89: 0.003263
90: -0.002640
91: -0.001716
92: 0.001821
93: -0.003002
94: -0.000306
95: 0.003186
96: -0.001615
97: 0.001524
98: 0.003364
99: -0.002062
```
</details><br/>

__Pytorch__
```python
""" In utils.py:do_detect() """

print(model.models[0][0].weight.data.storage()[0:100])
return
```
<details><summary>Pytorch Output</summary>

```
 0.002494276035577059
 -0.05781181901693344
 0.008488495834171772
 0.027733728289604187
 -0.1260443925857544
 0.01913182996213436
 0.09067513793706894
 -0.01978287287056446
 0.03458346053957939
 0.025900166481733322
 -0.05741793289780617
 0.027847010642290115
 0.05988399684429169
 -0.13511911034584045
 0.018076132982969284
 0.09771779179573059
 -0.05505229905247688
 0.013282321393489838
 0.024292904883623123
 -0.022654302418231964
 0.0045770504511892796
 0.019716612994670868
 -0.09117753058671951
 -0.005139993969351053
 0.053772397339344025
 -0.0036676565650850534
 -0.0001853857102105394
 0.0005600443691946566
 0.03491080924868584
 0.055543314665555954
 0.06160866469144821
 0.03578753396868706
 -0.04813152179121971
 -0.04790978506207466
 0.03329552710056305
 -0.08830180764198303
 -0.05267126485705376
 0.04684116691350937
 0.055535029619932175
 0.06602399051189423
 0.04059464856982231
 -0.06963779032230377
 -0.06726450473070145
 0.039733003824949265
 -0.09375892579555511
 -0.052297528833150864
 -0.04379674047231674
 -0.0457538478076458
 -0.025025997310876846
 -0.04407551884651184
 -0.07699237763881683
 -0.05740084499120712
 -0.009686418808996677
 -0.05727509781718254
 -0.01881682127714157
 0.1537521779537201
 -0.138362854719162
 -0.051714953035116196
 0.2643453776836395
 -0.0733807161450386
 -0.19294185936450958
 0.22541560232639313
 0.03976327180862427
 -0.17647702991962433
 0.18930558860301971
 -0.18341077864170074
 -0.0935848206281662
 0.38082441687583923
 -0.09357381612062454
 -0.3168611526489258
 0.2976643741130829
 0.06087712198495865
 -0.24723853170871735
 0.07735296338796616
 -0.07615843415260315
 -0.030188951641321182
 0.1314631998538971
 -0.018913179636001587
 -0.1209639459848404
 0.08420757949352264
 0.0471462644636631
 -0.11747846752405167
 -0.0024144472554326057
 -0.0015100453747436404
 0.0018667072290554643
 -0.002610392402857542
 -9.563190542394295e-05
 0.0030898943077772856
 -0.0014626922784373164
 0.0015505451010540128
 0.0032630902715027332
 -0.002639928599819541
 -0.0017157612601295114
 0.001820504548959434
 -0.0030017169192433357
 -0.00030643490026704967
 0.003186470363289118
 -0.0016146564157679677
 0.0015240473439916968
 0.0033637506421655416
[torch.cuda.FloatStorage of size 100]
```
</details><br>

These results are pretty weird, seeing as no one has ever encountered this (or at least submitted an issue). Could it be an endianness issue or something like that?

**Comment by 0phoff** [comment]
OK, should have tested this way sooner, but I finally found the problem!
Apparently pjreddie made a new weight-file version.

The header of my weightfile says it is on version 0.2.0, whilst the darknet19_448.conv.23 etc. are on version 0.1.0...
I'll try to find the changes pjreddie made to the files, so this repo can be updated to work with the newest version of weight files :medal_military:

**Comment by 0phoff** [comment]
The difference between v0.1.0 and v0.2.0 is that he uses a size_t variable to hold the number of images seen. (self.seen in the python code). This is rather annoying because, iirc, the size of a size_t variable is dependent on the stdlib version that is used! When parsing the file, we cannot check what version was used to create the file, because the weight file might have been created on a different computer, etc.

As a temporary solution, we might read the seen variable as a 8 byte variable, when the version is v0.2.0, as this is mostly the case! But this is really something that should be fixed by pjreddie in the actual darknet code, by eg. casting the size_t to a int64..

I am not closing this issue, as it is still a bug in the code, but now at least we know this problem is there and how it can be fixed upstream :clown_face:

**Comment by marvis** [comment]
@0phoff  Thanks for your patient work! The darknet version I used is https://github.com/marvis/darknet. Maybe helps for you.

**Comment by 0phoff** [comment]
Np.. and yeah, I believe you didn't pull in the last changes from pjreddie.. :)
